index/scorch/merge.go

```diff
@@ -129,8 +146,65 @@ func (s *Scorch) parseMergePlannerOptions() (*mergeplan.MergePlanOptions,
 	return &mergePlannerOptions, nil
 }
 
+type mergerKick struct {
+	moptions *mergeplan.MergePlanOptions
+	doneCh   chan struct{}
+}
+
+func (s *Scorch) RequestMerge(mo *mergeplan.MergePlanOptions) error {
+	// check whether force merge is already under processing
+	s.rootLock.Lock()
+	if atomic.LoadUint64(&s.stats.TotFileMergeForceOpsStarted) >
+		atomic.LoadUint64(&s.stats.TotFileMergeForceOpsCompleted) {
+		s.rootLock.Unlock()
+		return fmt.Errorf("compaction already in progress")
+	}
+	s.rootLock.Unlock()
+
+	go func() {
+		atomic.AddUint64(&s.stats.TotFileMergeForceOpsStarted, 1)
+		for {
+			msg := &mergerKick{moptions: mo, doneCh: make(chan struct{})}
+			// kick the merger
+			select {
+			case s.mergerKickCh <- msg:
+			case <-s.closeCh:
+				return
+			}
+			// stats before the merge cycle
+			totalNonePlans := atomic.LoadUint64(&s.stats.TotFileMergePlanNone)
+			totalOkPlans := atomic.LoadUint64(&s.stats.TotFileMergePlanOk)
+			totalBatchInro := atomic.LoadUint64(&s.stats.TotIntroducedSegmentsBatch)
+
+			// wait for the next merge loop completion
+			select {
+			case <-msg.doneCh:
+			case <-s.closeCh:
+				return
+			}
+			// with incoming mutations in place, bail out early than recursively
+			// applying the merge operations until the merge plan is met
+			// as that might turn out to be very resource intensive
+			if totalBatchInro != atomic.LoadUint64(&s.stats.TotIntroducedSegmentsBatch) {
+				atomic.AddUint64(&s.stats.TotFileMergeForceOpsSkipped, 1)
+				atomic.AddUint64(&s.stats.TotFileMergeForceOpsCompleted, 1)
+				return
+			}
+			// check whether merger has completed merging as per the given
+			// merge plan
+			if totalNonePlans != atomic.LoadUint64(&s.stats.TotFileMergePlanNone) &&
+				totalOkPlans == atomic.LoadUint64(&s.stats.TotFileMergePlanOk) {
+				atomic.AddUint64(&s.stats.TotFileMergeForceOpsCompleted, 1)
+				return
+			}
+		}
+	}()
+
+	return nil
+}
+
 func (s *Scorch) planMergeAtSnapshot(ourSnapshot *IndexSnapshot,
-	options *mergeplan.MergePlanOptions) error {
+	options *mergeplan.MergePlanOptions, forceCompact bool) error {
```

Comment:
one thought... thinking why not pass in the creator string as param?  The design "theory" (that's too fancy a word?) or the API "taste" there is that it'd be a more generalized API than a narrower "forceCompact" flag?